File: filepicker/src/main/java/droidninja/filepicker/adapters/SelectableAdapter.java
package droidninja.filepicker.adapters;

import android.support.v7.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import droidninja.filepicker.PickerManager;
import droidninja.filepicker.models.BaseFile;

public abstract class SelectableAdapter<VH extends RecyclerView.ViewHolder, T extends BaseFile>
    extends RecyclerView.Adapter<VH> implements Selectable<T> {

  private static final String TAG = SelectableAdapter.class.getSimpleName();
  private List<T> items;

  protected List<T> selectedPhotos;

  public SelectableAdapter(List<T> items, List<String> selectedPaths) {
    this.items = items;
    selectedPhotos = new ArrayList<>();

    addPathsToSelections(selectedPaths);
  }

  private void addPathsToSelections(List<String> selectedPaths) {
    if(selectedPaths==null)
      return;

    for (int index = 0; index < items.size(); index++) {
      for (int jindex = 0; jindex < selectedPaths.size(); jindex++) {
        if(items.get(index).getPath().equals(selectedPaths.get(jindex)))
        {
          selectedPhotos.add(items.get(index));
        }
      }
    }
  }


  /**
   * Indicates if the item at position where is selected
   *
   * @param photo Media of the item to check
   * @return true if the item is selected, false otherwise
   */
  @Override
  public boolean isSelected(T photo) {
    return selectedPhotos.contains(photo);
  }


  /**
   * Toggle the selection status of the item at a given position
   *
   * @param photo Media of the item to toggle the selection status for
   */
  @Override
  public void toggleSelection(T photo) {
    if (selectedPhotos.contains(photo)) {
      selectedPhotos.remove(photo);
    } else {
      selectedPhotos.add(photo);
    }
  }


  /**
   * Clear the selection status for all items
   */
  @Override
  public void clearSelection() {
    selectedPhotos.clear();
  }

  @Override
  public int getSelectedItemCount() {
    return selectedPhotos.size();
  }

  public void setData(List<T> items) {
    this.items = items;
  }

  public List<T> getItems()
  {
    return items;
  }

}

File: filepicker/src/main/java/droidninja/filepicker/adapters/SelectableAdapter.kt
package droidninja.filepicker.adapters

import android.net.Uri
import androidx.recyclerview.widget.RecyclerView

import java.util.ArrayList

import droidninja.filepicker.PickerManager
import droidninja.filepicker.models.BaseFile

abstract class SelectableAdapter<VH : RecyclerView.ViewHolder, T : BaseFile>(var items: List<T>, var selectedPaths: MutableList<Uri> = mutableListOf()) : RecyclerView.Adapter<VH>(), Selectable<T> {

    override val selectedItemCount: Int
        get() = selectedPaths.size
    /**
     * Indicates if the item at position where is selected
     *
     * @param photo Media of the item to check
     * @return true if the item is selected, false otherwise
     */
    override fun isSelected(item: T): Boolean {
        return selectedPaths.contains(item.path)
    }

    /**
     * Toggle the selection status of the item at a given position
     *
     * @param photo Media of the item to toggle the selection status for
     */
    override fun toggleSelection(item: T) {
        if (selectedPaths.contains(item.path)) {
            selectedPaths.remove(item.path)
        } else {
            selectedPaths.add(item.path)
        }
    }

    /**
     * Clear the selection status for all items
     */
    override fun clearSelection() {
        selectedPaths.clear()
        notifyDataSetChanged()
    }

    fun selectAll() {
        selectedPaths.clear()
        selectedPaths.addAll(items.map {
            it.path
        })
        notifyDataSetChanged()
    }

    fun setData(items: List<T>, selectedPaths: MutableList<Uri>) {
        this.items = items
        this.selectedPaths = selectedPaths
        notifyDataSetChanged()
    }

    companion object {

        private val TAG = "SelectableAdapter"
    }
}